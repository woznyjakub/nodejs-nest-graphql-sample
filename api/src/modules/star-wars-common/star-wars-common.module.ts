import { Module } from '@nestjs/common';

import { StarWarsCommonService } from './star-wars-common.service';

@Module({
  providers: [StarWarsCommonService],
  exports: [StarWarsCommonService],
})
export class StarWarsCommonModule {}
