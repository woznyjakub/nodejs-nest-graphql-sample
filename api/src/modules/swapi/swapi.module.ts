import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';

import { SwapiService } from './swapi.service';

@Module({
  imports: [HttpModule],
  providers: [SwapiService],
  exports: [SwapiService],
})
export class SwapiModule {}
