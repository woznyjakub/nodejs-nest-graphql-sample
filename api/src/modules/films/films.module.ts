import { Module } from '@nestjs/common';

import { StarWarsCommonModule } from '../star-wars-common/star-wars-common.module';
import { SwapiModule } from '../swapi/swapi.module';

import { FilmsResolver } from './films.resolver';
import { FilmsService } from './films.service';

@Module({
  imports: [SwapiModule, StarWarsCommonModule],
  providers: [FilmsResolver, FilmsService],
})
export class FilmsModule {}
