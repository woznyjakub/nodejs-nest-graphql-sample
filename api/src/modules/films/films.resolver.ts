import { Resolver, Query, Args, Int } from '@nestjs/graphql';

import { Film } from './entities/film.entity';
import { Films } from './entities/films.entity';
import { FilmsService } from './films.service';

@Resolver(() => Film)
export class FilmsResolver {
  constructor(private readonly vehiclesService: FilmsService) {}

  @Query(() => Films, { name: 'films' })
  findAll(@Args('page', { type: () => Int }) page: number): Promise<Films> {
    return this.vehiclesService.findAll(page);
  }

  @Query(() => Film, { name: 'film' })
  findOne(@Args('id', { type: () => Int }) id: number): Promise<Film> {
    return this.vehiclesService.findOne(id);
  }
}
