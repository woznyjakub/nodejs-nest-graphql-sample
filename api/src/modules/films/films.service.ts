import { Injectable } from '@nestjs/common';

import { StarWarsCommonService } from '../star-wars-common/star-wars-common.service';
import { FilmResponse } from '../swapi/dto/film';
import { SwapiService } from '../swapi/swapi.service';

import { Film } from './entities/film.entity';
import { Films } from './entities/films.entity';

@Injectable()
export class FilmsService {
  constructor(
    private readonly swapiService: SwapiService,
    private readonly swCommonService: StarWarsCommonService,
  ) {}

  async findAll(page: number): Promise<Films> {
    const swapiData = await this.swCommonService.fetchWithCaching(
      this.swCommonService.getFilmsCacheKey(page),
      () => this.swapiService.getFilms(page),
    );

    return {
      count: swapiData.count,
      next: swapiData.next,
      previous: swapiData.previous,
      results: swapiData.results.map(this.mapSingleFilm),
    };
  }

  async findOne(id: number): Promise<Film> {
    const swapiData = await this.swCommonService.fetchWithCaching(
      this.swCommonService.getFilmCacheKey(id),
      () => this.swapiService.getFilm(id),
    );

    return this.mapSingleFilm(swapiData);
  }

  private mapSingleFilm(film: FilmResponse): Film {
    return {
      title: film.title,
      episodeId: film.episode_id,
      openingCrawl: film.opening_crawl,
      director: film.director,
      producer: film.producer,
      releaseDate: film.release_date,
      characters: film.characters,
      planets: film.planets,
      starships: film.starships,
      vehicles: film.vehicles,
      species: film.species,
      url: film.url,
    };
  }

  async getUniqueWordsCount(): Promise<[string, number][]> {
    const { results: films } = await this.swCommonService.fetchWithCaching(
      this.swCommonService.getFilmsCacheKey(1),
      () => this.swapiService.getFilms(),
    );

    const wordCounts: Map<string, number> = new Map();

    films.forEach((film) => {
      const words = film.opening_crawl
        .split(/\W/)
        .filter((word) => word.trim() !== '')
        .map((word) => word.toLowerCase());

      words.forEach((word) => {
        const currentCount = wordCounts.get(word) || 0;
        wordCounts.set(word, currentCount + 1);
      });
    });

    return Array.from(wordCounts).sort((a, b) => b[1] - a[1]);
  }

  async getMostFrequentCharacterInFilms(): Promise<string[]> {
    const films = await this.swCommonService.getAllPages(
      this.swCommonService.getFilmsCacheKey,
      this.swapiService.getFilms,
    );

    const combinedOpeningCrawl = films
      .map((film) => film.opening_crawl)
      .join(' ')
      .toLowerCase();

    const people = await this.swCommonService.getAllPages(
      this.swCommonService.getPeopleCacheKey,
      this.swapiService.getPeople,
    );

    const characterOccurrences: Map<string, number> = new Map();

    people.forEach((person) => {
      // eslint-disable-next-line
      const regex = new RegExp(`\\b${person.name.toLowerCase()}\\b`, 'g'); // Dopasowanie całego słowa
      const matchCount = (combinedOpeningCrawl.match(regex) || []).length;

      if (matchCount > 0) {
        characterOccurrences.set(person.name, matchCount);
      }
    });

    const maxOccurrences = Math.max(...characterOccurrences.values());

    return Array.from(characterOccurrences.entries())
      .filter(([, count]) => count === maxOccurrences)
      .map(([name]) => name);
  }
}
