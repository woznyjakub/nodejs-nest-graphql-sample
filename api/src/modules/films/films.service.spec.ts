import { Test, TestingModule } from '@nestjs/testing';

import { filmResponseMock, filmsResponseMock } from '../../test/mocks/swapi';
import { StarWarsCommonService } from '../star-wars-common/star-wars-common.service';
import { SwapiService } from '../swapi/swapi.service';

import { FilmsService } from './films.service';

describe('FilmsService', () => {
  let filmsService: FilmsService;

  const mockFetchWithCaching = vi.fn();

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        {
          provide: StarWarsCommonService,
          useValue: {
            fetchWithCaching: mockFetchWithCaching,
          },
        },
        FilmsService,
        {
          provide: SwapiService,
          useValue: {
            getFilm: vi.fn(),
            getFilms: vi.fn(),
          },
        },
      ],
    }).compile();

    filmsService = module.get<FilmsService>(FilmsService);
  });

  describe('single film data', () => {
    it('should return mapped data', async () => {
      mockFetchWithCaching.mockReturnValue(filmResponseMock);

      const result = await filmsService.findOne(1);

      expect(result).toMatchSnapshot();
    });
  });

  describe('multiple films data', () => {
    it('should return mapped data when', async () => {
      mockFetchWithCaching.mockReturnValue(filmsResponseMock);

      const result = await filmsService.findAll(1);

      expect(result).toMatchSnapshot();
    });
  });
});
