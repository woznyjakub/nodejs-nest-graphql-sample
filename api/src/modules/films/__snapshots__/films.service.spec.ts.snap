// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilmsService > multiple films data > should return mapped data when 1`] = `
{
  "count": 6,
  "next": null,
  "previous": null,
  "results": [
    {
      "characters": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/3/",
        "https://swapi.dev/api/people/4/",
        "https://swapi.dev/api/people/5/",
        "https://swapi.dev/api/people/6/",
        "https://swapi.dev/api/people/7/",
        "https://swapi.dev/api/people/8/",
        "https://swapi.dev/api/people/9/",
        "https://swapi.dev/api/people/10/",
        "https://swapi.dev/api/people/12/",
        "https://swapi.dev/api/people/13/",
        "https://swapi.dev/api/people/14/",
        "https://swapi.dev/api/people/15/",
        "https://swapi.dev/api/people/16/",
        "https://swapi.dev/api/people/18/",
        "https://swapi.dev/api/people/19/",
        "https://swapi.dev/api/people/81/",
      ],
      "director": "George Lucas",
      "episodeId": 4,
      "openingCrawl": "It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy....",
      "planets": [
        "https://swapi.dev/api/planets/1/",
        "https://swapi.dev/api/planets/2/",
        "https://swapi.dev/api/planets/3/",
      ],
      "producer": "Gary Kurtz, Rick McCallum",
      "releaseDate": "1977-05-25",
      "species": [
        "https://swapi.dev/api/species/1/",
        "https://swapi.dev/api/species/2/",
        "https://swapi.dev/api/species/3/",
        "https://swapi.dev/api/species/4/",
        "https://swapi.dev/api/species/5/",
      ],
      "starships": [
        "https://swapi.dev/api/starships/2/",
        "https://swapi.dev/api/starships/3/",
        "https://swapi.dev/api/starships/5/",
        "https://swapi.dev/api/starships/9/",
        "https://swapi.dev/api/starships/10/",
        "https://swapi.dev/api/starships/11/",
        "https://swapi.dev/api/starships/12/",
        "https://swapi.dev/api/starships/13/",
      ],
      "title": "A New Hope",
      "url": "https://swapi.dev/api/films/1/",
      "vehicles": [
        "https://swapi.dev/api/vehicles/4/",
        "https://swapi.dev/api/vehicles/6/",
        "https://swapi.dev/api/vehicles/7/",
        "https://swapi.dev/api/vehicles/8/",
      ],
    },
    {
      "characters": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/3/",
        "https://swapi.dev/api/people/4/",
        "https://swapi.dev/api/people/5/",
        "https://swapi.dev/api/people/10/",
        "https://swapi.dev/api/people/13/",
        "https://swapi.dev/api/people/14/",
        "https://swapi.dev/api/people/18/",
        "https://swapi.dev/api/people/20/",
        "https://swapi.dev/api/people/21/",
        "https://swapi.dev/api/people/22/",
        "https://swapi.dev/api/people/23/",
        "https://swapi.dev/api/people/24/",
        "https://swapi.dev/api/people/25/",
        "https://swapi.dev/api/people/26/",
      ],
      "director": "Irvin Kershner",
      "episodeId": 5,
      "openingCrawl": "It is a dark time for the
Rebellion. Although the Death
Star has been destroyed,
Imperial troops have driven the
Rebel forces from their hidden
base and pursued them across
the galaxy.

Evading the dreaded Imperial
Starfleet, a group of freedom
fighters led by Luke Skywalker
has established a new secret
base on the remote ice world
of Hoth.

The evil lord Darth Vader,
obsessed with finding young
Skywalker, has dispatched
thousands of remote probes into
the far reaches of space....",
      "planets": [
        "https://swapi.dev/api/planets/4/",
        "https://swapi.dev/api/planets/5/",
        "https://swapi.dev/api/planets/6/",
        "https://swapi.dev/api/planets/27/",
      ],
      "producer": "Gary Kurtz, Rick McCallum",
      "releaseDate": "1980-05-17",
      "species": [
        "https://swapi.dev/api/species/1/",
        "https://swapi.dev/api/species/2/",
        "https://swapi.dev/api/species/3/",
        "https://swapi.dev/api/species/6/",
        "https://swapi.dev/api/species/7/",
      ],
      "starships": [
        "https://swapi.dev/api/starships/3/",
        "https://swapi.dev/api/starships/10/",
        "https://swapi.dev/api/starships/11/",
        "https://swapi.dev/api/starships/12/",
        "https://swapi.dev/api/starships/15/",
        "https://swapi.dev/api/starships/17/",
        "https://swapi.dev/api/starships/21/",
        "https://swapi.dev/api/starships/22/",
        "https://swapi.dev/api/starships/23/",
      ],
      "title": "The Empire Strikes Back",
      "url": "https://swapi.dev/api/films/2/",
      "vehicles": [
        "https://swapi.dev/api/vehicles/8/",
        "https://swapi.dev/api/vehicles/14/",
        "https://swapi.dev/api/vehicles/16/",
        "https://swapi.dev/api/vehicles/18/",
        "https://swapi.dev/api/vehicles/19/",
        "https://swapi.dev/api/vehicles/20/",
      ],
    },
    {
      "characters": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/3/",
        "https://swapi.dev/api/people/4/",
        "https://swapi.dev/api/people/5/",
        "https://swapi.dev/api/people/10/",
        "https://swapi.dev/api/people/13/",
        "https://swapi.dev/api/people/14/",
        "https://swapi.dev/api/people/16/",
        "https://swapi.dev/api/people/18/",
        "https://swapi.dev/api/people/20/",
        "https://swapi.dev/api/people/21/",
        "https://swapi.dev/api/people/22/",
        "https://swapi.dev/api/people/25/",
        "https://swapi.dev/api/people/27/",
        "https://swapi.dev/api/people/28/",
        "https://swapi.dev/api/people/29/",
        "https://swapi.dev/api/people/30/",
        "https://swapi.dev/api/people/31/",
        "https://swapi.dev/api/people/45/",
      ],
      "director": "Richard Marquand",
      "episodeId": 6,
      "openingCrawl": "Luke Skywalker has returned to
his home planet of Tatooine in
an attempt to rescue his
friend Han Solo from the
clutches of the vile gangster
Jabba the Hutt.

Little does Luke know that the
GALACTIC EMPIRE has secretly
begun construction on a new
armored space station even
more powerful than the first
dreaded Death Star.

When completed, this ultimate
weapon will spell certain doom
for the small band of rebels
struggling to restore freedom
to the galaxy...",
      "planets": [
        "https://swapi.dev/api/planets/1/",
        "https://swapi.dev/api/planets/5/",
        "https://swapi.dev/api/planets/7/",
        "https://swapi.dev/api/planets/8/",
        "https://swapi.dev/api/planets/9/",
      ],
      "producer": "Howard G. Kazanjian, George Lucas, Rick McCallum",
      "releaseDate": "1983-05-25",
      "species": [
        "https://swapi.dev/api/species/1/",
        "https://swapi.dev/api/species/2/",
        "https://swapi.dev/api/species/3/",
        "https://swapi.dev/api/species/5/",
        "https://swapi.dev/api/species/6/",
        "https://swapi.dev/api/species/8/",
        "https://swapi.dev/api/species/9/",
        "https://swapi.dev/api/species/10/",
        "https://swapi.dev/api/species/15/",
      ],
      "starships": [
        "https://swapi.dev/api/starships/2/",
        "https://swapi.dev/api/starships/3/",
        "https://swapi.dev/api/starships/10/",
        "https://swapi.dev/api/starships/11/",
        "https://swapi.dev/api/starships/12/",
        "https://swapi.dev/api/starships/15/",
        "https://swapi.dev/api/starships/17/",
        "https://swapi.dev/api/starships/22/",
        "https://swapi.dev/api/starships/23/",
        "https://swapi.dev/api/starships/27/",
        "https://swapi.dev/api/starships/28/",
        "https://swapi.dev/api/starships/29/",
      ],
      "title": "Return of the Jedi",
      "url": "https://swapi.dev/api/films/3/",
      "vehicles": [
        "https://swapi.dev/api/vehicles/8/",
        "https://swapi.dev/api/vehicles/16/",
        "https://swapi.dev/api/vehicles/18/",
        "https://swapi.dev/api/vehicles/19/",
        "https://swapi.dev/api/vehicles/24/",
        "https://swapi.dev/api/vehicles/25/",
        "https://swapi.dev/api/vehicles/26/",
        "https://swapi.dev/api/vehicles/30/",
      ],
    },
    {
      "characters": [
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/3/",
        "https://swapi.dev/api/people/10/",
        "https://swapi.dev/api/people/11/",
        "https://swapi.dev/api/people/16/",
        "https://swapi.dev/api/people/20/",
        "https://swapi.dev/api/people/21/",
        "https://swapi.dev/api/people/32/",
        "https://swapi.dev/api/people/33/",
        "https://swapi.dev/api/people/34/",
        "https://swapi.dev/api/people/35/",
        "https://swapi.dev/api/people/36/",
        "https://swapi.dev/api/people/37/",
        "https://swapi.dev/api/people/38/",
        "https://swapi.dev/api/people/39/",
        "https://swapi.dev/api/people/40/",
        "https://swapi.dev/api/people/41/",
        "https://swapi.dev/api/people/42/",
        "https://swapi.dev/api/people/43/",
        "https://swapi.dev/api/people/44/",
        "https://swapi.dev/api/people/46/",
        "https://swapi.dev/api/people/47/",
        "https://swapi.dev/api/people/48/",
        "https://swapi.dev/api/people/49/",
        "https://swapi.dev/api/people/50/",
        "https://swapi.dev/api/people/51/",
        "https://swapi.dev/api/people/52/",
        "https://swapi.dev/api/people/53/",
        "https://swapi.dev/api/people/54/",
        "https://swapi.dev/api/people/55/",
        "https://swapi.dev/api/people/56/",
        "https://swapi.dev/api/people/57/",
        "https://swapi.dev/api/people/58/",
        "https://swapi.dev/api/people/59/",
      ],
      "director": "George Lucas",
      "episodeId": 1,
      "openingCrawl": "Turmoil has engulfed the
Galactic Republic. The taxation
of trade routes to outlying star
systems is in dispute.

Hoping to resolve the matter
with a blockade of deadly
battleships, the greedy Trade
Federation has stopped all
shipping to the small planet
of Naboo.

While the Congress of the
Republic endlessly debates
this alarming chain of events,
the Supreme Chancellor has
secretly dispatched two Jedi
Knights, the guardians of
peace and justice in the
galaxy, to settle the conflict....",
      "planets": [
        "https://swapi.dev/api/planets/1/",
        "https://swapi.dev/api/planets/8/",
        "https://swapi.dev/api/planets/9/",
      ],
      "producer": "Rick McCallum",
      "releaseDate": "1999-05-19",
      "species": [
        "https://swapi.dev/api/species/1/",
        "https://swapi.dev/api/species/2/",
        "https://swapi.dev/api/species/6/",
        "https://swapi.dev/api/species/11/",
        "https://swapi.dev/api/species/12/",
        "https://swapi.dev/api/species/13/",
        "https://swapi.dev/api/species/14/",
        "https://swapi.dev/api/species/15/",
        "https://swapi.dev/api/species/16/",
        "https://swapi.dev/api/species/17/",
        "https://swapi.dev/api/species/18/",
        "https://swapi.dev/api/species/19/",
        "https://swapi.dev/api/species/20/",
        "https://swapi.dev/api/species/21/",
        "https://swapi.dev/api/species/22/",
        "https://swapi.dev/api/species/23/",
        "https://swapi.dev/api/species/24/",
        "https://swapi.dev/api/species/25/",
        "https://swapi.dev/api/species/26/",
        "https://swapi.dev/api/species/27/",
      ],
      "starships": [
        "https://swapi.dev/api/starships/31/",
        "https://swapi.dev/api/starships/32/",
        "https://swapi.dev/api/starships/39/",
        "https://swapi.dev/api/starships/40/",
        "https://swapi.dev/api/starships/41/",
      ],
      "title": "The Phantom Menace",
      "url": "https://swapi.dev/api/films/4/",
      "vehicles": [
        "https://swapi.dev/api/vehicles/33/",
        "https://swapi.dev/api/vehicles/34/",
        "https://swapi.dev/api/vehicles/35/",
        "https://swapi.dev/api/vehicles/36/",
        "https://swapi.dev/api/vehicles/37/",
        "https://swapi.dev/api/vehicles/38/",
        "https://swapi.dev/api/vehicles/42/",
      ],
    },
    {
      "characters": [
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/3/",
        "https://swapi.dev/api/people/6/",
        "https://swapi.dev/api/people/7/",
        "https://swapi.dev/api/people/10/",
        "https://swapi.dev/api/people/11/",
        "https://swapi.dev/api/people/20/",
        "https://swapi.dev/api/people/21/",
        "https://swapi.dev/api/people/22/",
        "https://swapi.dev/api/people/33/",
        "https://swapi.dev/api/people/35/",
        "https://swapi.dev/api/people/36/",
        "https://swapi.dev/api/people/40/",
        "https://swapi.dev/api/people/43/",
        "https://swapi.dev/api/people/46/",
        "https://swapi.dev/api/people/51/",
        "https://swapi.dev/api/people/52/",
        "https://swapi.dev/api/people/53/",
        "https://swapi.dev/api/people/58/",
        "https://swapi.dev/api/people/59/",
        "https://swapi.dev/api/people/60/",
        "https://swapi.dev/api/people/61/",
        "https://swapi.dev/api/people/62/",
        "https://swapi.dev/api/people/63/",
        "https://swapi.dev/api/people/64/",
        "https://swapi.dev/api/people/65/",
        "https://swapi.dev/api/people/66/",
        "https://swapi.dev/api/people/67/",
        "https://swapi.dev/api/people/68/",
        "https://swapi.dev/api/people/69/",
        "https://swapi.dev/api/people/70/",
        "https://swapi.dev/api/people/71/",
        "https://swapi.dev/api/people/72/",
        "https://swapi.dev/api/people/73/",
        "https://swapi.dev/api/people/74/",
        "https://swapi.dev/api/people/75/",
        "https://swapi.dev/api/people/76/",
        "https://swapi.dev/api/people/77/",
        "https://swapi.dev/api/people/78/",
        "https://swapi.dev/api/people/82/",
      ],
      "director": "George Lucas",
      "episodeId": 2,
      "openingCrawl": "There is unrest in the Galactic
Senate. Several thousand solar
systems have declared their
intentions to leave the Republic.

This separatist movement,
under the leadership of the
mysterious Count Dooku, has
made it difficult for the limited
number of Jedi Knights to maintain 
peace and order in the galaxy.

Senator Amidala, the former
Queen of Naboo, is returning
to the Galactic Senate to vote
on the critical issue of creating
an ARMY OF THE REPUBLIC
to assist the overwhelmed
Jedi....",
      "planets": [
        "https://swapi.dev/api/planets/1/",
        "https://swapi.dev/api/planets/8/",
        "https://swapi.dev/api/planets/9/",
        "https://swapi.dev/api/planets/10/",
        "https://swapi.dev/api/planets/11/",
      ],
      "producer": "Rick McCallum",
      "releaseDate": "2002-05-16",
      "species": [
        "https://swapi.dev/api/species/1/",
        "https://swapi.dev/api/species/2/",
        "https://swapi.dev/api/species/6/",
        "https://swapi.dev/api/species/12/",
        "https://swapi.dev/api/species/13/",
        "https://swapi.dev/api/species/15/",
        "https://swapi.dev/api/species/28/",
        "https://swapi.dev/api/species/29/",
        "https://swapi.dev/api/species/30/",
        "https://swapi.dev/api/species/31/",
        "https://swapi.dev/api/species/32/",
        "https://swapi.dev/api/species/33/",
        "https://swapi.dev/api/species/34/",
        "https://swapi.dev/api/species/35/",
      ],
      "starships": [
        "https://swapi.dev/api/starships/21/",
        "https://swapi.dev/api/starships/32/",
        "https://swapi.dev/api/starships/39/",
        "https://swapi.dev/api/starships/43/",
        "https://swapi.dev/api/starships/47/",
        "https://swapi.dev/api/starships/48/",
        "https://swapi.dev/api/starships/49/",
        "https://swapi.dev/api/starships/52/",
        "https://swapi.dev/api/starships/58/",
      ],
      "title": "Attack of the Clones",
      "url": "https://swapi.dev/api/films/5/",
      "vehicles": [
        "https://swapi.dev/api/vehicles/4/",
        "https://swapi.dev/api/vehicles/44/",
        "https://swapi.dev/api/vehicles/45/",
        "https://swapi.dev/api/vehicles/46/",
        "https://swapi.dev/api/vehicles/50/",
        "https://swapi.dev/api/vehicles/51/",
        "https://swapi.dev/api/vehicles/53/",
        "https://swapi.dev/api/vehicles/54/",
        "https://swapi.dev/api/vehicles/55/",
        "https://swapi.dev/api/vehicles/56/",
        "https://swapi.dev/api/vehicles/57/",
      ],
    },
    {
      "characters": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/3/",
        "https://swapi.dev/api/people/4/",
        "https://swapi.dev/api/people/5/",
        "https://swapi.dev/api/people/6/",
        "https://swapi.dev/api/people/7/",
        "https://swapi.dev/api/people/10/",
        "https://swapi.dev/api/people/11/",
        "https://swapi.dev/api/people/12/",
        "https://swapi.dev/api/people/13/",
        "https://swapi.dev/api/people/20/",
        "https://swapi.dev/api/people/21/",
        "https://swapi.dev/api/people/33/",
        "https://swapi.dev/api/people/35/",
        "https://swapi.dev/api/people/46/",
        "https://swapi.dev/api/people/51/",
        "https://swapi.dev/api/people/52/",
        "https://swapi.dev/api/people/53/",
        "https://swapi.dev/api/people/54/",
        "https://swapi.dev/api/people/55/",
        "https://swapi.dev/api/people/56/",
        "https://swapi.dev/api/people/58/",
        "https://swapi.dev/api/people/63/",
        "https://swapi.dev/api/people/64/",
        "https://swapi.dev/api/people/67/",
        "https://swapi.dev/api/people/68/",
        "https://swapi.dev/api/people/75/",
        "https://swapi.dev/api/people/78/",
        "https://swapi.dev/api/people/79/",
        "https://swapi.dev/api/people/80/",
        "https://swapi.dev/api/people/81/",
        "https://swapi.dev/api/people/82/",
        "https://swapi.dev/api/people/83/",
      ],
      "director": "George Lucas",
      "episodeId": 3,
      "openingCrawl": "War! The Republic is crumbling
under attacks by the ruthless
Sith Lord, Count Dooku.
There are heroes on both sides.
Evil is everywhere.

In a stunning move, the
fiendish droid leader, General
Grievous, has swept into the
Republic capital and kidnapped
Chancellor Palpatine, leader of
the Galactic Senate.

As the Separatist Droid Army
attempts to flee the besieged
capital with their valuable
hostage, two Jedi Knights lead a
desperate mission to rescue the
captive Chancellor....",
      "planets": [
        "https://swapi.dev/api/planets/1/",
        "https://swapi.dev/api/planets/2/",
        "https://swapi.dev/api/planets/5/",
        "https://swapi.dev/api/planets/8/",
        "https://swapi.dev/api/planets/9/",
        "https://swapi.dev/api/planets/12/",
        "https://swapi.dev/api/planets/13/",
        "https://swapi.dev/api/planets/14/",
        "https://swapi.dev/api/planets/15/",
        "https://swapi.dev/api/planets/16/",
        "https://swapi.dev/api/planets/17/",
        "https://swapi.dev/api/planets/18/",
        "https://swapi.dev/api/planets/19/",
      ],
      "producer": "Rick McCallum",
      "releaseDate": "2005-05-19",
      "species": [
        "https://swapi.dev/api/species/1/",
        "https://swapi.dev/api/species/2/",
        "https://swapi.dev/api/species/3/",
        "https://swapi.dev/api/species/6/",
        "https://swapi.dev/api/species/15/",
        "https://swapi.dev/api/species/19/",
        "https://swapi.dev/api/species/20/",
        "https://swapi.dev/api/species/23/",
        "https://swapi.dev/api/species/24/",
        "https://swapi.dev/api/species/25/",
        "https://swapi.dev/api/species/26/",
        "https://swapi.dev/api/species/27/",
        "https://swapi.dev/api/species/28/",
        "https://swapi.dev/api/species/29/",
        "https://swapi.dev/api/species/30/",
        "https://swapi.dev/api/species/33/",
        "https://swapi.dev/api/species/34/",
        "https://swapi.dev/api/species/35/",
        "https://swapi.dev/api/species/36/",
        "https://swapi.dev/api/species/37/",
      ],
      "starships": [
        "https://swapi.dev/api/starships/2/",
        "https://swapi.dev/api/starships/32/",
        "https://swapi.dev/api/starships/48/",
        "https://swapi.dev/api/starships/59/",
        "https://swapi.dev/api/starships/61/",
        "https://swapi.dev/api/starships/63/",
        "https://swapi.dev/api/starships/64/",
        "https://swapi.dev/api/starships/65/",
        "https://swapi.dev/api/starships/66/",
        "https://swapi.dev/api/starships/68/",
        "https://swapi.dev/api/starships/74/",
        "https://swapi.dev/api/starships/75/",
      ],
      "title": "Revenge of the Sith",
      "url": "https://swapi.dev/api/films/6/",
      "vehicles": [
        "https://swapi.dev/api/vehicles/33/",
        "https://swapi.dev/api/vehicles/50/",
        "https://swapi.dev/api/vehicles/53/",
        "https://swapi.dev/api/vehicles/56/",
        "https://swapi.dev/api/vehicles/60/",
        "https://swapi.dev/api/vehicles/62/",
        "https://swapi.dev/api/vehicles/67/",
        "https://swapi.dev/api/vehicles/69/",
        "https://swapi.dev/api/vehicles/70/",
        "https://swapi.dev/api/vehicles/71/",
        "https://swapi.dev/api/vehicles/72/",
        "https://swapi.dev/api/vehicles/73/",
        "https://swapi.dev/api/vehicles/76/",
      ],
    },
  ],
}
`;

exports[`FilmsService > single film data > should return mapped data 1`] = `
{
  "characters": [
    "https://swapi.dev/api/people/1/",
    "https://swapi.dev/api/people/2/",
    "https://swapi.dev/api/people/3/",
    "https://swapi.dev/api/people/4/",
    "https://swapi.dev/api/people/5/",
    "https://swapi.dev/api/people/6/",
    "https://swapi.dev/api/people/7/",
    "https://swapi.dev/api/people/8/",
    "https://swapi.dev/api/people/9/",
    "https://swapi.dev/api/people/10/",
    "https://swapi.dev/api/people/12/",
    "https://swapi.dev/api/people/13/",
    "https://swapi.dev/api/people/14/",
    "https://swapi.dev/api/people/15/",
    "https://swapi.dev/api/people/16/",
    "https://swapi.dev/api/people/18/",
    "https://swapi.dev/api/people/19/",
    "https://swapi.dev/api/people/81/",
  ],
  "director": "George Lucas",
  "episodeId": 4,
  "openingCrawl": "It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy....",
  "planets": [
    "https://swapi.dev/api/planets/1/",
    "https://swapi.dev/api/planets/2/",
    "https://swapi.dev/api/planets/3/",
  ],
  "producer": "Gary Kurtz, Rick McCallum",
  "releaseDate": "1977-05-25",
  "species": [
    "https://swapi.dev/api/species/1/",
    "https://swapi.dev/api/species/2/",
    "https://swapi.dev/api/species/3/",
    "https://swapi.dev/api/species/4/",
    "https://swapi.dev/api/species/5/",
  ],
  "starships": [
    "https://swapi.dev/api/starships/2/",
    "https://swapi.dev/api/starships/3/",
    "https://swapi.dev/api/starships/5/",
    "https://swapi.dev/api/starships/9/",
    "https://swapi.dev/api/starships/10/",
    "https://swapi.dev/api/starships/11/",
    "https://swapi.dev/api/starships/12/",
    "https://swapi.dev/api/starships/13/",
  ],
  "title": "A New Hope",
  "url": "https://swapi.dev/api/films/1/",
  "vehicles": [
    "https://swapi.dev/api/vehicles/4/",
    "https://swapi.dev/api/vehicles/6/",
    "https://swapi.dev/api/vehicles/7/",
    "https://swapi.dev/api/vehicles/8/",
  ],
}
`;
